source/extensions/filters/network/local_ratelimit/local_ratelimit.cc

```diff
@@ -10,26 +10,91 @@ namespace Extensions {
 namespace NetworkFilters {
 namespace LocalRateLimitFilter {
 
+SINGLETON_MANAGER_REGISTRATION(shared_local_ratelimit);
+
+SharedRateLimitSingleton::~SharedRateLimitSingleton() {
+  // Validate that all limiters were properly cleaned up when Configs are deleted.
+  ASSERT(limiters_.empty());
+}
+
+std::pair<LocalRateLimiterImplSharedPtr, const SharedRateLimitSingleton::Key*>
+SharedRateLimitSingleton::get(
+    const envoy::extensions::filters::network::local_ratelimit::v3::LocalRateLimit& proto_config,
+    std::function<LocalRateLimiterImplSharedPtr()> create_fn) {
+  ASSERT_IS_MAIN_OR_TEST_THREAD();
+
+  if (proto_config.share_key().empty()) {
+    // If there is no key, never share.
+    return {create_fn(), nullptr};
+  }
+
+  LocalRateLimiterImplSharedPtr rate_limiter;
+  const Key key{proto_config.share_key(), proto_config.token_bucket()};
+  auto it = limiters_.find(key);
+  if (it == limiters_.end()) {
+    ENVOY_LOG(debug, fmt::format("LocalRateLimit for share_key '{}' is creating a new token bucket "
+                                 "due to no match found",
+                                 proto_config.share_key()));
+    rate_limiter = create_fn();
+    it = limiters_.emplace(key, rate_limiter).first;
```

Comment:
Following up on our previous discussion here (whether it is thread safe, and that this is part of the config):
Isn't this singleton table accessed by the worker-threads as it is part of the shared config?
I guess the naming "Config" is a bit confusing me, especially the comment [here](https://github.com/envoyproxy/envoy/blob/cefb8f866e393ed87f54c6c030b5b82d9c11037a/source/extensions/filters/network/local_ratelimit/local_ratelimit.h#L60-L63).

I guess that the assertion above holds if the main thread is always the one that creates this Config and just passes it along to other threads, so as long as the `shared_bucket_registry_` is only accessed in the c'tor (or by methods that aren't called by worker threads) then it should be ok.

Reply:
Yep, it works because the Config constructor does the lookup, and it is always constructed on the main thread.

Reply:
If that’s the case then I think that adding the assertion should ensure that it won’t be misused.